src/test/kotlin/com/safehill/kcrypto/CipherTests.kt

```diff
@@ -1,112 +1,133 @@
+                secret = secret,
+                instant = Instant.now()
+            )
+            delay(1.seconds)
+            val (code2, _) = SafehillOTP(digits = 6, validDuration = 1.seconds).generateCode(
+                secret = secret,
+                instant = Instant.now()
+            )
+            val (code3, _) = SafehillOTP(digits = 6, validDuration = 1.seconds).generateCode(
+                secret = secret,
+                instant = Instant.now() - 1.seconds.toJavaDuration()
+            )
+            assertNotEquals(code1, code2)
+            assertEquals(code1, code3)
+        }
+    }
 
-        val (code8, valid8) = SafehillCypher.generateOTPCode(
-            secret = secret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        Thread.sleep(max(0, valid8 - 100))
-        val (code9, valid9) = SafehillCypher.generateOTPCode(
+    @Test
+    fun `test generated code is different after valid time has elapsed`() {
+        val secret = SafehillCypher.generateRandomIV()
+        val safehillOTP = SafehillOTP(digits = 6, validDuration = 2.seconds)
+        val instant = Instant.now()
+        val (code1, validity) = safehillOTP.generateCode(
             secret = secret,
-            digits = 6,
-            timeStepInSeconds = 2
+            instant = instant
         )
-        assertEquals(code8, code9)
-        Thread.sleep(valid9 + 1)
-        val (code10, _) = SafehillCypher.generateOTPCode(
+        val secondInstant = Instant.now()
+        val (code2, _) = safehillOTP.generateCode(
             secret = secret,
-            digits = 6,
-            timeStepInSeconds = 2
+            instant = secondInstant + validity.toJavaDuration() + 1.milliseconds.toJavaDuration()
         )
-        assertNotEquals(code9, code10)
+        assertNotEquals(code1, code2)
+    }
+    @Test
+    fun `test code is same when we move to second slot but verify in the first slot`(){
+        runBlocking {
+            val secret = SafehillCypher.generateRandomIV()
+            val safehillOTP = SafehillOTP(digits = 6, validDuration = 2.seconds)
+            val (code1, ) = safehillOTP.generateCode(secret = secret, instant = Instant.now())
+            delay(2.seconds)
+            val (code2, _) = safehillOTP.generateCode(secret = secret, instant = Instant.now() - 2.seconds.toJavaDuration())
+            assertEquals(code1, code2)
+        }
+    }
 
-        val newSecret = SafehillCypher.generateRandomIV()
-        val (code11, valid11) = SafehillCypher.generateOTPCode(
-            secret = newSecret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        Thread.sleep(max(0, valid11 - 100))
-        val (code12, _) = SafehillCypher.generateOTPCode(
-            secret = newSecret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        assertEquals(code11, code12)
+    @Test
+    fun `test generated code is different after valid time`() {
```

Comment:
What's the difference between this and `test generated code is different after valid time has elapsed`?

Reply:
Its objectively the same thing. In one case we are delaying the time till we reach the second slot. In another, we are programmatically increasing the Instant to the second slot.